File: Python/arrays_and_hashing/twosum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        numIndexDict = {}
        for i, n in enumerate(nums):
            diff = target - n
            if diff in numIndexDict:
                return [numIndexDict[diff], i]
            numIndexDict[n] = i

File: JavaScript/arrays_and_hashing/twosum.js
// Optimized Solution
// Hash table to check the difference
// Time Complexity: O(n)
// Space Complexity: O(n)

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  let numIndexDict = {};
  for (let i = 0; i < nums.length; i++) {
    let diff = target - nums[i];
    if (diff in numIndexDict) {
      return [numIndexDict[diff], i];
    }

    numIndexDict[nums[i]] = i;
  }
};
